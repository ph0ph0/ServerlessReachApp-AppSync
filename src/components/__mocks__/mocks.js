
import ListRecipes from '../../graphql/queries/ListRecipes'

const mocks = [
    {
        request: {
            query: ListRecipes
        }, 
        result: {
            data: {
                listRecipes: {
                    items: [
                        {
                            name: "Tuna Rolll",
                            id: "1",
                            instructions: ["Mix Tuna"],
                            ingredients: ["Tuna"]
                        }  
                    ]                  
                }
            }
        }
    }
]

const data = {
    listRecipes: {
        items: [
            {
                id: "1",
                name: "Tuna Roll",
                instructions: ["Mix Tuna"],
                ingredients: ["Tuna"]
            }  
        ]                  
    }
}


export default mocks
export {data}