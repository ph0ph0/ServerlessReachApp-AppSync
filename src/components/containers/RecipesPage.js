import React, { useState } from 'react'
import {useQuery, useSubscription} from '@apollo/react-hooks'
import ListRecipes from '../../graphql/queries/ListRecipes'
import NewRecipeSubscription from '../../graphql/subscriptions/NewRecipeSubscription'
import RecipeList from '../RecipeList'

const Recipes = () => {

    //We don't have to set state because anytime the component renders, the query will be called.
    //It will try and get the data from our cache first if it can, and if not it will send off to the network.
    // const [recipesState, setRecipeState] = useState(null)

    const { data, loading, error} = useQuery(ListRecipes)
    const {loading: subLoading, error: subError} = useSubscription(NewRecipeSubscription)

    console.log(`data: ${JSON.stringify(data)}`)

    //Will need some sort of state to 
    //hold the data that comes from the query. Then, everytime the 
    //subscription fires, we append it to the state.

    //We want to do the first fetch when the component mounts and then update thereafter. Actually, 
    //useQuery fires when the component first mounts and so we dont need to put this into the
    //useEffect statement. We also need to provide an unmount listener call

    if (loading) return <p data-testid = "Loading">Loading...</p>
    if (error) {
        console.log(`Error getting ListRecipes: ${error}`)
        return <p>Error!</p>
    }

    return (
        <div data-testid = "recipeComponent">
            <RecipeList  data = {data}/>
            {subLoading && <p>Loading</p>}
            {subError && <p>Error</p>}
        </div>
        
    )
}

export default Recipes