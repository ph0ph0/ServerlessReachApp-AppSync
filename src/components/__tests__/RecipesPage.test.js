//Should render DONE
//Should get data from query
//Should display loading when query fetching
//Should show error if query fails
//Should display query results on page
//Should link subscription on mount
//Should unlink sub on unmount
//Should update query results on page when results change

import React from 'react'
import RecipesPage from '../containers/RecipesPage'

import {render, cleanup, waitForElement } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { MockedProvider } from '@apollo/react-testing'
import mocks from '../__mocks__/mocks'

afterEach(cleanup)

it("gets data from ListRecipess", async () => {

    const { getByTestId } = render(
        <MockedProvider mocks = {mocks} addTypename = {false}>
                <RecipesPage />
        </MockedProvider>
    )

    expect(getByTestId("Loading")).toHaveTextContent("Loading...")

    const resolvedDataDiv = await waitForElement(() => (getByTestId("recipeComponent")))
    
    expect(resolvedDataDiv).toBeInTheDocument()


})

it("Recipe page renders", () => {
    const { asFragment } = render(
        <MockedProvider mocks = {mocks} addTypename = {false}>
                <RecipesPage />
        </MockedProvider>
    )
    expect(asFragment()).toMatchSnapshot()
})