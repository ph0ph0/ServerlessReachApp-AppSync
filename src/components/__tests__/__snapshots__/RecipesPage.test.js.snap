// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe page renders 1`] = `
<DocumentFragment>
  <p
    data-testid="Loading"
  >
    Loading...
  </p>
</DocumentFragment>
`;
